### PR TITLE
PWN-3686 - UI Kit Buttons

### PR DESCRIPTION
**Pull request by kuznetsov**
## Jira Ticket

https://p2pvalidator.atlassian.net/browse/PWN-3686
https://www.figma.com/file/13KdUCrfZ6VPBdfT6sLVbi/Android-UI-Kit?node-id=167%3A2243

## Description of Work

UI Kit buttons implementation.

### Styles structure

```xml
<style name="UiKit.Atoms.Button" parent="Widget.MaterialComponents.Button.UnelevatedButton" />

<style name="UiKit.Atoms.Button.Large" />

<style name="UiKit.Atoms.Button.Medium" />

<style name="UiKit.Atoms.Button.Small" />

<style name="UiKit.Atoms.Button.Large.Text" />

<style name="UiKit.Atoms.Button.Medium.Text" />

<style name="UiKit.Atoms.Button.Small.Text" />
```

### Example

```xml
<org.p2p.uikit.atoms.UiKitButton
    android:id="@+id/buttonLarge"
    style="@style/UiKit.Atoms.Button.Large"
    android:layout_width="wrap_content"
    android:layout_height="@dimen/ui_kit_button_large_height"
    android:backgroundTint="@color/bg_night"
    android:textColor="@color/text_lime"
    app:icon="@drawable/ic_arrow_back_24px"
    app:iconGravity="textStart"
    app:iconTint="@color/icons_lime" />
```

## Screenshots

Screenshot 2022-06-16 at 17 00 46

### CODE REVIEW

File: ui-kit/src/main/res/values/styles_buttons.xml [hunk 1]
```diff
@@ -0,0 +1,45 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+
+    <style name="UiKit.Atom.Button" parent="Widget.MaterialComponents.Button.UnelevatedButton">
```
Comment (kuznetsov) [inline ui-kit/src/main/res/values/styles_buttons.xml:4]: Maybe UiKit.**Atoms**.Button will be better?

File: ui-kit/src/main/kotlin/org/p2p/uikit/atoms/UiKitButton.kt [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+                circularProgressDrawable.start()
+            } else {
+                icon = currentIcon
+                circularProgressDrawable.stop()
+            }
+            field = loading
+        }
+
+    private val drawableCallback = object : Drawable.Callback {
+        override fun unscheduleDrawable(who: Drawable, what: Runnable) {}
```
Comment (gslevinkov) [inline ui-kit/src/main/kotlin/org/p2p/uikit/atoms/UiKitButton.kt:39]:
```suggestion
        override fun unscheduleDrawable(who: Drawable, what: Runnable) = NoOp
```

File: ui-kit/src/main/kotlin/org/p2p/uikit/atoms/UiKitButton.kt [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+        override fun invalidateDrawable(who: Drawable) {
+            this@UiKitButton.invalidate()
+        }
+
+        override fun scheduleDrawable(who: Drawable, what: Runnable, `when`: Long) {}
```
Comment (gslevinkov) [inline ui-kit/src/main/kotlin/org/p2p/uikit/atoms/UiKitButton.kt:45]:
```suggestion
        override fun scheduleDrawable(who: Drawable, what: Runnable, `when`: Long) = NoOp
```

File: ui-kit/src/main/kotlin/org/p2p/uikit/atoms/UiKitButton.kt [hunk 4]
```diff
@@ -0,0 +1,67 @@
+package org.p2p.uikit.atoms
+
+import android.content.Context
+import android.graphics.drawable.Drawable
+import android.util.AttributeSet
+import androidx.swiperefreshlayout.widget.CircularProgressDrawable
+import com.google.android.material.button.MaterialButton
+import org.p2p.uikit.R
+import org.p2p.uikit.utils.toPx
```
Comment (gslevinkov) [inline ui-kit/src/main/kotlin/org/p2p/uikit/atoms/UiKitButton.kt:9]:
```suggestion
import org.p2p.uikit.utils.toPx
import org.p2p.wallet.utils.NoOp
```

File: ui-kit/src/main/kotlin/org/p2p/uikit/atoms/UiKitButton.kt [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    attrs: AttributeSet? = null,
+    defStyleAttr: Int = 0
+) : MaterialButton(context, attrs, defStyleAttr) {
+
+    private var currentIcon: Drawable? = null
+    private val circularProgressDrawable = CircularProgressDrawable(context)
+
+    var isLoading: Boolean = false
+        set(loading) {
+            if (loading) {
```
Comment (gslevinkov) [inline ui-kit/src/main/kotlin/org/p2p/uikit/atoms/UiKitButton.kt:26]: if's can be extracted to methods like `renderLoadingButton`

File: ui-kit/src/main/res/values/styles_buttons.xml [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+        <item name="android:paddingTop">0dp</item>
+        <item name="android:paddingBottom">0dp</item>
+        <item name="android:insetTop">0dp</item>
+        <item name="android:insetBottom">0dp</item>
+        <item name="iconPadding">10dp</item>
+    </style>
+
+    <style name="UiKit.Atom.Button.Large">
+        <item name="cornerRadius">12dp</item>
```
[the remark this answers is not in the document]
Comment (gslevinkov) [reply ui-kit/src/main/res/values/styles_buttons.xml:13]: neh, im okay with hardcode

**Review by gslevinkov — APPROVED** [review]

**Comment by gslevinkov** [comment]
no assignee

File: ui-kit/src/main/res/values/styles_buttons.xml (hunk 6 above)
[the remark this answers is not in the document]
Comment (kabramovp2p) [reply ui-kit/src/main/res/values/styles_buttons.xml:13]: im ok also

**Review by kabramovp2p — APPROVED** [review]

**Pull request merged**